source/iopipe/json/formatter.d

```diff
@@ -0,0 +1,466 @@
+        final switch (state) with (State)
+        {
+        case Begin, Member, Value:
+            break;
+        case First, End:
+            throw new JSONIopipeException(
+                "beginString not allowed in current state");
+        }
+
+        if (quoteChar != '"' && quoteChar != '\'')
+            throw new JSONIopipeException("Invalid quote character");
+
+        currentQuoteChar = quoteChar;
+        putStr((&quoteChar)[0 .. 1]);
+    }
+
+    // will automatically escape string data as needed.
+    void addStringData(bool validate = true, bool addEscapes = true, T)(T value) {
+        static if (validate) {
+            // Validate that the string doesn't contain invalid characters
+            foreach(char c; value) {
+                if (c < 0x20 && c != '\t' && c != '\n' && c != '\r') {
+                    throw new JSONIopipeException(format("Invalid control character \\u%04X in string", cast(int)c));
+                }
+            }
+        }
+    
+        static if (addEscapes) {
+            auto escaped = value.substitute!(jsonEscapeSubstitutions!()).to!string;
+            putStr(escaped);
+        } else {
+            putStr(value);
+        }
+    }
+
+
+    void endString() {
+        putStr((&currentQuoteChar)[0 .. 1]);
+
+        if (stackLen == 0)
+            state = State.End;
+        else
+        // If it's at the end of an array or an object,
+        // endArray() and endObject() will set the state appropriately
+            state = State.Member;
+    }
+
+    // allow numeric types, or strings that are validated to be JSON numbers
+    void addNumericData(T)(T value, string formatStr = "%s") {
+        
+        final switch (state) with (State)
+        {
+        case Begin, Member, Value:
+            break;
+        case First, End:
+            throw new JSONIopipeException(
+                "addNumericData not allowed in current state");
+        }
+        
+        static if (is(T == string)) {
+            if (!isValidJSONNumber(value)) {
+                throw new JSONIopipeException(format("Invalid JSON number: %s", value));
+            }
+            putStr(value);
+        } else {
+            static assert(isNumeric!T, "addNumericData requires a numeric type");
+            formattedWrite(&putStr, formatStr, value);
+        }
+
+        if (stackLen == 0)
+            state = State.End;
+        else
+            state = State.Member;
+    }
+
+    // null, true, false, inf, etc.
+    void addKeywordValue(KeywordValue value) {
+        final switch (state) with (State)
+        {
+        case Begin, Member, Value:
+            break;
+        case First, End:
+            throw new JSONIopipeException(
+                "addKeywordValue not allowed in current state");
+        }
+
+        putStr(value);
+
+        if (stackLen == 0)
+            state = State.End;
+        else
+            state = State.Member;
+    }
+
+    void addWhitespace(T)(T data) {
```

Comment:
Having second thoughts on the `addWhitespace` member.

I'm thinking that the whitespace is one of the major features of the formatter (essentially the indentation and spacing).

However, we likely need to support all whitespace output at all places where it makes sense (between tokens). But this can be added later.

In any case, can we remove this function for now?